package/samplers/hebo/sampler.py

```diff
@@ -18,19 +24,136 @@
+            A :class:`~optuna.samplers.BaseSampler` instance that is used for independent
+            sampling. The parameters not contained in the relative search space are sampled
+            by this sampler. If :obj:`None` is specified, :class:`~optuna.samplers.RandomSampler`
+            is used as the default.
+    """  # NOQA
+
+    def __init__(
+        self,
+        search_space: dict[str, BaseDistribution] | None = None,
+        *,
+        seed: int | None = None,
+        constant_liar: bool = False,
+        independent_sampler: BaseSampler | None = None,
+    ) -> None:
+        super().__init__(search_space, seed)
+        if search_space is not None and not constant_liar:
+            self._hebo = HEBO(self._convert_to_hebo_design_space(search_space), scramble_seed=seed)
+        else:
+            self._hebo = None
+        self._intersection_search_space = IntersectionSearchSpace()
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._is_fallback_inevitable = False
+        self._constant_liar = constant_liar
+        self._rng = np.random.default_rng(seed)
+
+    def _sample_relative_define_and_run(
         self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
     ) -> dict[str, float]:
-        params_pd = self._hebo.suggest()
+        return {
+            name: row.iloc[0]
+            for name, row in self._hebo.suggest().items()
+            if name in search_space.keys()
+        }
 
-        params = {}
-        for name in search_space.keys():
-            params[name] = params_pd[name].to_numpy()[0]
-        return params
+    def _sample_relative_stateless(
+        self, study: Study, trial: FrozenTrial, search_space: dict[str, BaseDistribution]
+    ) -> dict[str, float]:
+        if self._constant_liar:
+            target_states = [TrialState.COMPLETE, TrialState.RUNNING]
+        else:
+            target_states = [TrialState.COMPLETE]
+
+        use_cache = not self._constant_liar
+        trials = study._get_trials(deepcopy=False, states=target_states, use_cache=use_cache)
+        if len([t for t in trials if t.state == TrialState.COMPLETE]) < 1:
+            # note: The backend HEBO implementation uses Sobol sampling here.
+            # This sampler does not call `hebo.suggest()` here because
+            # Optuna needs to know search space by running the first trial in Define-by-Run.
+            self._is_fallback_inevitable = True
+            return {}
+        else:
+            self._is_fallback_inevitable = False
+
+        # Assume that the back-end HEBO implementation aims to minimize.
+        if study.direction == StudyDirection.MINIMIZE:
+            worst_value = max(t.value for t in trials if t.state == TrialState.COMPLETE)
+        else:
+            worst_value = min(t.value for t in trials if t.state == TrialState.COMPLETE)
+        sign = 1 if study.direction == StudyDirection.MINIMIZE else -1
```

Comment:
Probably, `self._is_independent_sample_necessary` is easier to understand.

```suggestion
        is_complete = np.array([t.state == TrialState.COMPLETE for t in trials])
        if not np.any(is_complete):
            # note: The backend HEBO implementation uses Sobol sampling here.
            # This sampler does not call `hebo.suggest()` here because
            # Optuna needs to know search space by running the first trial in Define-by-Run.
            self._is_independent_sample_necessary = True
            return {}
        else:
            self._is_independent_sample_necessary = False

        # Assume that the back-end HEBO implementation aims to minimize.
        values = np.array([t.value if t.state == TrialState.COMPLETE else np.nan for t in trials])
        worst_value = np.nanmax(values) if study.direction == StudyDirection.MINIMIZE else np.nanmin(values)
        sign = 1 if study.direction == StudyDirection.MINIMIZE else -1
```